apollo-router/src/cache/mod.rs

```diff
@@ -113,7 +113,11 @@ where
     }
 
     pub(crate) async fn insert(&self, key: K, value: V) {
-        self.storage.insert(key, value.clone()).await;
+        // Let's see if we can help waiters make progress.
+        if let Some(sender) = self.wait_map.lock().await.remove(&key) {
+            let _ = sender.send(value.clone());
+        }
```

Comment:
cc @Geal @garypen because I honestly don't understand all the intricacies of the cache, all I know is the test used to fail, and not it passes.

Reply:
I think the problem you are resolving is that there may be waiters who are not notified that a value has been inserted. The bit I don't understand is that, if there are waiters, then something should be "getting" the cached value and that should resolve at some point.

Logically, it's probably not safe to insert() a value to the cache if there are waiters for that key, so I think that aspect of your change is correct, but I don't like the fact that waiters should be resolved at some point and maybe insert should be ordered behind the existing get?

I think it needs input from @Geal .

Reply:
right, we paired on this and it turns out to be a corner case we hit because the test runs in the single threaded flavor. the actual observable behavior is that if it ever happens, one(1) client will need to send the apq and the query once more, which is totally ok.

Moreover this allows us to /not/ send the value like that. the core issue stems from the fact that the cache assumes the first person who tries to `get` is going to `insert`, which doesn't apply to APQ.